add modulemdmerge: Merge several modules.yaml files into one

g-v-egidy opened the pull request.

FrostyX commented:
Hello @g-v-egidy,
thank you for the PR!

I was actually about to write exactly this feature for PR #4, so I will just call this script.

I will try to review it soon.

g-v-egidy commented on modulemdmerge/README.md:
```diff
@@ -0,0 +1,41 @@
+# modulemdmerge
```
no problem, I renamed it to modulemd-merge.

Although the pylint you wanted now complains "Module name "modulemd-merge" doesn't conform to snake_case naming style". I think python naming would prefer "modulemd_merge".

g-v-egidy replied:
If you actually wanted "modulemd_merge", this is fine by me too.

g-v-egidy commented on README.md:
```diff
@@ -13,3 +13,5 @@ dir2module - Generates a module YAML definition based on essential module
     the module are found in a specified directory or a text file containing
     their list.
+
+modulemdmerge - merge several modules.yaml files into one
```
I rebased my commit

g-v-egidy commented:
> May I be interested in what is going to be your personal usage?

My usecase is creating a local mirror of Centos 8 that does keep files that were removed from the mirror source. 

This is to allow to downgrade packages in case of a problem with an updated package. Getting the expired binaries can be quite a challenge otherwise. For finding out a package update caused some problem, I want to use the exact binaries that were installed before.

> Also, can you please try to run `pylint ./modulemdmerge.py` and address those issues? There is just a few of them, the script already looks quite good but we can still polish it a little.

I addressed most of them, except these ones:

- Line too long (106/100)

Fixed line lengths of 80 or 100 chars are a bit of a legacy thing, with widescreen monitors being in use everywhere now. Using the available screen width improves readability in contrast to many broken up lines that add more length.

Also these warnings are also for example in dir2module.py

- Module name "modulemd-merge" doesn't conform to snake_case naming style 

see my other comment

- Missing module docstring

adapt to the style without docstrings as common in the rest of modulemd-tools.

- Import "from gi.repository import Modulemd" should be placed at the top of the module

this is a bogus warning by pylint, the import from gi should not be done before the gi.require_version

- Use % formatting in logging functions and pass the % parameters as arguments

string formatting is a dark corner of python. I like this quote about it "The string interpolation syntax in Python is basically Guido van Rossum's Dorian Grey portrait in the closet. The more clean and streamlined the rest of his language gets, the more confusing and Perl-like Python string interpolation gets" ( https://twitter.com/mcclure111/status/1278518340273745921 )

That the % encoding is hard-coded into the logging system is really bad. So I just pre-format the strings.

So I think keeping the remaining pylint warnings is justified. Please elaborate if you think otherwise.

FrostyX commented:
Thank you very much for the PR @g-v-egidy and for those changes.
I am merging now.

g-v-egidy commented:
Thank you very much.

To make using this more easy, could you add some RedHat-internal karma to https://bugzilla.redhat.com/show_bug.cgi?id=1860401
Thanks.
